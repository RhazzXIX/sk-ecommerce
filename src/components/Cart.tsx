import {  useEffect, useState } from "react";
import del from "../../public/trash-can-outline.svg";
import Image from "next/image";

const Cart = ({
  products,
  deleteProduct,
  showLogin,
}: {
  products: product[];
  deleteProduct: (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
  showLogin: () => void;
}) => {
  const [totalCost, setTotalCost] = useState(0);
  const getTotalCost = () => {
    let addedCost = 0;
    products.forEach((product) => {
      if (product.qty) {
        const productCost = product.price * product.qty;
        addedCost += productCost;
      }
    });
    setTotalCost(addedCost);
  };

  useEffect(() => {
    getTotalCost();
  }, [products]);

  return (
    <section id="cart-page">
      <header>
        <h4>Cart</h4>
        <h5>Items:</h5>
        <h5>Costs:</h5>
      </header>
      {products.length !== 0 ? (
        <ul>
          {products.map((product) => {
            if (!product.qty) return null;
            return (
              <li key={product.id}>
                <h4>{product.title}</h4>
                <Image
                  src={product.images[0]}
                  alt={product.title}
                  width={0}
                  height={0}
                />
                <p>
                  $ {product.price}.00 x {product.qty}
                </p>
                <p>$ {product.price * product.qty}.00</p>
                <button onClick={deleteProduct} data-id={product.id}>
                  <Image
                    src={del}
                    alt="delete"
                    data-id={product.id}
                    width={10}
                    height={10}
                  />
                </button>
              </li>
            );
          })}
        </ul>
      ) : (
        <p>No items on the cart</p>
      )}
      <footer>
        <h4>Total Cost: $ {totalCost}.00</h4>
        <button onClick={showLogin}>Login to Pay</button>
      </footer>
    </section>
  );
};

export default Cart;
