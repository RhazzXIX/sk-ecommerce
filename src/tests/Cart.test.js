import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import Cart from "../components/Cart";

const user = userEvent.setup();

const products = [];

const delProd = () => {
  products.pop();
};

let loginIsShown = false;

const showLogin = () => {
  loginIsShown = true;
};

const product = {
  title: "Black Motorbike",
  id: "214",
  price: 569,
  images: ["https://i.dummyjson.com/data/products/91/1.jpg"],
  qty: 2,
  description:
    "Engine Type:Wet sump, Single Cylinder, Four Stroke, Two Valves, Air Cooled with SOHC (Single Over Head Cam) Chain Drive Bore & Stroke:47.0 x 49.5 MM",
};

const product2 = {
  title: "iPhone X",
  id: "456",
  price: 899,
  images: ["https://i.dummyjson.com/data/products/2/1.jpg"],
  qty: 1,
  description:
    "SIM-Free, Model A19211 6.5-inch Super Retina HD display with OLED technology A12 Bionic chip",
};

describe("Cart component", () => {
  beforeEach(() => {
    products.push(product);
    loginIsShown = false;
  });
  it("Gets rendered", () => {
    const { container } = render(<Cart products={[]} />);

    expect(container).toMatchSnapshot();
  });

  it(`Shows items, added to the cart`, () => {
    const { rerender } = render(<Cart products={[]} deleteProduct={delProd} />);
    expect(screen.getByText("No items on the cart")).toBeInTheDocument();
    expect(screen.queryByRole("list")).not.toBeInTheDocument();
    rerender(<Cart products={products} deleteProduct={delProd} />);
    expect(screen.queryByText("No items on the cart")).not.toBeInTheDocument();
    expect(screen.getByRole("list")).toBeInTheDocument();
    expect(screen.getByRole("listitem")).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Black Motorbike" })
    ).toBeInTheDocument();
  });

  it("Can delete products", async () => {
    const { rerender } = render(
      <Cart products={products} deleteProduct={delProd} />
    );
    const delBtn = screen.getByAltText("delete");
    expect(screen.getByRole("listitem")).toBeInTheDocument();

    await user.click(delBtn);
    rerender(<Cart products={products} deleteProduct={delProd} />);
    expect(screen.queryByRole("listitem")).not.toBeInTheDocument();
  });

  it("Shows the total cost of all the products", () => {
    const { rerender } = render(
      <Cart products={products} deleteProduct={delProd} />
    );
    expect(
      screen.queryByRole("heading", { name: /total/i }).textContent
    ).toMatch("Total Cost: $ 1138.00");

    const addedProducts = [...products, product2];
    rerender(<Cart products={addedProducts} deleteProduct={delProd} />);
    expect(
      screen.queryByRole("heading", { name: /total/i }).textContent
    ).toMatch("Total Cost: $ 2037.00");
  });

  it("Lets shows the login section when the user press the login button", async () => {
    render(
      <Cart products={[]} deleteProduct={delProd} showLogin={showLogin} />
    );
    const loginBtn = screen.getByRole("button", { name: /Login/i });

    await user.click(loginBtn);

    expect(loginIsShown).toBeTruthy();
  });

  afterEach(() => products.splice(0));
});
