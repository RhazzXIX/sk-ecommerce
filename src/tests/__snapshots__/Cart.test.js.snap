// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component Gets rendered 1`] = `
<div>
  <section
    id="cart-page"
  >
    <header>
      <h4>
        Cart
      </h4>
      <h5>
        Items:
      </h5>
      <h5>
        Costs:
      </h5>
    </header>
    <p>
      No items on the cart
    </p>
    <footer>
      <h4>
        Total Cost: $ 
        0
        .00
      </h4>
      <button>
        Login to Pay
      </button>
    </footer>
  </section>
</div>
`;
